lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -4417,24 +4469,32 @@ ConformanceChecker::resolveWitnessViaLookup(ValueDecl *requirement) {
         auto protoAccessScope = proto->getFormalAccessScope(DC);
         bool protoForcesAccess =
           requiredAccessScope.hasEqualDeclContextWith(protoAccessScope);
-        auto diagKind = protoForcesAccess
-                          ? diag::witness_not_accessible_proto
-                          : diag::witness_not_accessible_type;
         bool isSetter = (check.Kind == CheckKind::AccessOfSetter);
 
         auto &diags = DC->getASTContext().Diags;
-        diags.diagnose(getLocForDiagnosingWitness(conformance, witness),
+        auto *decl = dyn_cast<AbstractFunctionDecl>(witness);
+        auto *ctor = dyn_cast<ConstructorDecl>(witness);
+
+        auto diagKind = protoForcesAccess
+            ? (ctor && ctor->isSynthesized()) ? diag::implicit_init_witness_not_accessible_proto : diag::witness_not_accessible_proto
+            : diag::witness_not_accessible_type;
+
+        SourceLoc diagLoc = getLocForDiagnosingWitness(conformance, witness);
+        diags.diagnose(diagLoc,
                        diagKind, getProtocolRequirementKind(requirement),
                        witness, isSetter, requiredAccess,
                        protoAccessScope.accessLevelForDiagnostics(),
                        proto);
 
-        auto *decl = dyn_cast<AbstractFunctionDecl>(witness);
-        if (decl && decl->isSynthesized())
+        if (ctor && ctor->isSynthesized()) {
+          diagnoseImplicitInitWitnessFixAccessLevel(diags, ctor, requiredAccess, conformance, diagLoc);
```

Comment:
I expect `isSynthesized` to return true for both memberwise and so-called default initializers:
```
// 'C' has an implicit default initializer.
class C {}
```

The default init in non-final class case is diagnosed separately (`witness_initializer_not_required`). I don’t think we want to complain again in that particular case. But if you want to cover the default init in *final* class case, I would leave this condition as is and add an early return in `diagnoseImplicitInitWitnessFixAccessLevel`.